Improve AWS guidelines on exception handling

willthames opened the pull request:
##### SUMMARY
Better document what exceptions to handle, when and why.
Describe how to handle client auth exceptions, and that
AWSRetry retries on `XYZNotFound` exceptions.


##### ISSUE TYPE
 - Docs Pull Request

##### COMPONENT NAME
modules/cloud/amazon guidelines

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel d72eb08902) last updated 2017/09/22 16:02:40 (GMT +1000)
  config file = None
  configured module search path = [u'/Users/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/will/src/ansible/lib/ansible
  executable location = /Users/will/src/ansible/bin/ansible
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

willthames commented:
cc @mikedlr

s-hertel commented on lib/ansible/modules/cloud/amazon/GUIDELINES.md:
```diff
@@ -142,7 +142,8 @@ else:
 exception handling like in boto.  Instead, an `AuthFailure` exception will be thrown when you use
-'connection'. See exception handling.
+'connection'. To ensure that authorization, parameter validation and permissions errors are all
+caught, you should catch `ClientError` and `BotoCoreError` exceptions with every boto3 connection call.
```
Do we know when BotoCoreError occurs rather than ClientError? I think I've only experience ClientError when providing invalid credentials.

willthames replied:
BotoCoreError covers ParamValidationError and NoCredentialsError

```
In [13]: inspect.getclasstree(inspect.getmro(botocore.exceptions.NoCredentialsError))
Out[13]: 
[(object, ()),
 [(BaseException, (object,)),
  [(Exception, (BaseException,)),
   [(botocore.exceptions.BotoCoreError, (Exception,)),
    [(botocore.exceptions.NoCredentialsError,
      (botocore.exceptions.BotoCoreError,))]]]]]
```

NoCredentialsError is what happens if you call any API method (e.g. client.describe_instances) without the credentials being properly passed (i.e. passing no values, rather than passing incorrect values)

With no default profile, the latter suffices to generate a NoCredentialsError:

```
import botocore
client = boto3.session.Session().client('ec2', region_name='ap-southeast-2')
client.describe_instances()
```

s-hertel commented on lib/ansible/modules/cloud/amazon/GUIDELINES.md:
````diff
@@ -240,36 +245,45 @@ Boto3 provides lots of useful information when an exception is thrown so pass th
                      **camel_dict_to_snake_dict(e.response))
 ```
 
+Note: we use `str(e)` rather than `e.message` as the latter doesn't
````
Should we really be using to_native()?

willthames replied:
I don't know - in what circumstances would `str(e)` cause a problem? (`str(e)` appears to work for both python2 and 3)

willthames replied:
This question also applies my second commit that I've just pushed to handle `ProfileNotFound`

s-hertel commented on lib/ansible/modules/cloud/amazon/GUIDELINES.md:
```diff
@@ -389,21 +433,6 @@ Ansible format, this function will convert the keys to snake_case.
 Converts a an Ansible list of filters to a boto3 friendly list of dicts.  This is useful for any
 boto3 `_facts` modules.
 
-### boto_exception
```
Since we still have modules using boto and some of the exception handling probably could be improved in places I feel like this section shouldn't be deleted yet.

willthames replied:
Ah, I thought `boto_exception` was less widely used than it appears to be, I've removed this part of the change.

s-hertel commented on lib/ansible/modules/cloud/amazon/GUIDELINES.md:
````diff
@@ -309,6 +326,36 @@ def describe_some_resource(client, module):
+the resource doesn't exist and not retry, we need:
+
+```
+@AWSRetry.exponential_backoff(tries=5, delay=5)
````
`tries` -> `retries` - do you mind fixing the one above too?

willthames replied:
Good catch, thank you

willthames commented:
Removed `ProfileNotFound` guidelines again in favour of #31921

s-hertel commented on lib/ansible/modules/cloud/amazon/GUIDELINES.md:
````diff
@@ -240,36 +245,45 @@ Boto3 provides lots of useful information when an exception is thrown so pass th
 
 ```python
-# Import ClientError from botocore
+from ansible.module_utils.ec2 import HAS_BOTO3
````
Since some of these examples are using AnsibleAWSModule, can you add something about not needing to import or check for HAS_BOTO3 if using it?

willthames replied:
Done, thanks

Commit:
Improve exception handling information

Better document what exceptions to handle, when and why.
Describe how to handle client auth exceptions, and that
AWSRetry retries on `XYZNotFound` exceptions.
